cupy/padding/pad.py

```diff
@@ -1,321 +1,788 @@
+            )
+            new_right_pad = right_pad - period
+        else:
+            # Chunk matches pad area
+            pad_area = _slice_at_axis(slice(-right_pad, None), axis)
+        padded[pad_area] = left_chunk
+
+    return new_left_pad, new_right_pad
+
+
+def _as_pairs(x, ndim, as_index=False):
+    """Broadcast `x` to an array with the shape (`ndim`, 2).
+
+    A helper function for `pad` that prepares and validates arguments like
+    `pad_width` for iteration in pairs.
 
     Args:
-        array (array-like): Input array of rank N.
-        pad_width (int or array-like): Number of values padded
-            to the edges of each axis.
-            ((before_1, after_1), ... (before_N, after_N)) uniquely pad widths
-            for each axis.
-            ((before, after),) yields same before and after pad for each axis.
-            (pad,) or int is a shortcut for before = after = pad width for all
-            axes.
-            You cannot specify ``cupy.ndarray`` .
-        mode (str):
-            'constant'
-                Pads with a constant values.
-            'edge'
-                Pads with the edge values of array.
-            'reflect'
-                Pads with the reflection of the vector mirrored on the first
-                and last values of the vector along each axis.
-        constant_values (int or array-like): Used in
-            ``constant``.
-            The values are padded for each axis.
-            ((before_1, after_1), ... (before_N, after_N)) uniquely pad
-            constants for each axis.
-            ((before, after),) yields same before and after constants for each
-            axis.
-            (constant,) or int is a shortcut for before = after = constant for
-            all axes.
-            Default is 0. You cannot specify ``cupy.ndarray`` .
-        reflect_type : {'even', 'odd'}, optional
-            Used in 'reflect', and 'symmetric'.  The 'even' style is the
-            default with an unaltered reflection around the edge value.  For
-            the 'odd' style, the extented part of the array is created by
-            subtracting the reflected values from two times the edge value.
+      x({None, scalar, array-like}): The object to broadcast to the shape
+          (`ndim`, 2).
+      ndim(int): Number of pairs the broadcasted `x` will have.
+      as_index(bool, optional, optional): If `x` is not None, try to round each
+          element of `x` to an integer (dtype `cupy.intp`) and ensure every
+          element is positive. (Default value = False)
 
     Returns:
-        cupy.ndarray:
-        Padded array of rank equal to ``array`` with shape increased according
-        to ``pad_width`` .
+      nested iterables, shape (`ndim`, 2): The broadcasted version of `x`.
+    """
+    if x is None:
+        # Pass through None as a special case, otherwise cupy.round(x) fails
+        # with an AttributeError
+        return ((None, None),) * ndim
+
+    x = numpy.array(x)
```

Comment:
Could you explain why is it need to convert to `numpy.ndarray`? Is it typo of `cupy.asarray(x)`?

Reply:
It is not a typo.  This `_as_pairs` function is basically just a utility that expands a user specified `pad_width` like to `2`, to its full form with a pre/post pad value for each axis (e.g.  `[(2, 2), (2, 2)]` in 2D).  This function uses `np.broadcast_to` to do this, but the returned values are a standard python list.  Given that the numpy arrays are just internal to this utility there is no reason to incur the overhead of transfer to/from the GPU.